src/core/public/context/context_service.ts

```diff
@@ -0,0 +1,115 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ContextContainer, ContextContainerImplementation } from './context';
+import { CoreContext } from '../core_system';
+
+interface StartDeps {
+  pluginDependencies: ReadonlyMap<string, string[]>;
+}
+
+/** @internal */
+export class ContextService {
+  constructor(private readonly core: CoreContext) {}
+
+  public setup({ pluginDependencies }: StartDeps): ContextSetup {
+    return {
+      createContextContainer: <
+        TContext extends {},
+        THandlerReturn,
+        THandlerParameters extends any[] = []
+      >() =>
+        new ContextContainerImplementation<TContext, THandlerReturn, THandlerParameters>(
+          pluginDependencies,
+          this.core.coreId
+        ),
+    };
+  }
+}
+
+/**
+ * {@inheritdoc ContextContainer}
+ *
+ * @example
+ * Say we're creating a plugin for rendering visualizations that allows new rendering methods to be registered. If we
+ * want to offer context to these rendering methods, we can leverage the ContextService to manage these contexts.
+ * ```ts
+ * export interface VizRenderContext {
+ *   core: {
+ *     i18n: I18nStart;
+ *     uiSettings: UISettingsClientContract;
+ *   }
+ *   [contextName: string]: unknown;
+ * }
+ *
+ * export type VizRenderer = (context: VizRenderContext, domElement: HTMLElement) => () => void;
+ *
+ * class VizRenderingPlugin {
+ *   private readonly vizRenderers = new Map<string, ((domElement: HTMLElement) => () => void)>();
+ *
+ *   setup(core) {
+ *     this.contextContainer = core.createContextContainer<
+ *       VizRenderContext,
+ *       ReturnType<VizRenderer>,
+ *       [HTMLElement]
+ *     >();
+ *
+ *     return {
+ *       registerContext: this.contextContainer.registerContext,
+ *       registerVizRenderer: (plugin: string, renderMethod: string, renderer: VizTypeRenderer) =>
+ *         this.vizRenderers.set(renderMethod, this.contextContainer.createHandler(plugin, renderer)),
+ *     };
+ *   }
+ *
+ *   start(core) {
+ *     // Register the core context available to all renderers. Use the VizRendererContext's pluginId as the first arg.
+ *     this.contextContainer.registerContext('viz_rendering', 'core', () => ({
+ *       i18n: core.i18n,
+ *       uiSettings: core.uiSettings
+ *     }));
+ *
+ *     return {
+ *       registerContext: this.contextContainer.registerContext,
+ *
+ *       // The handler can now be called directly with only an `HTMLElement` and will automaticallly
+ *       // have the `context` argument supplied.
+ *       renderVizualization: (renderMethod: string, domElement: HTMLElement) => {
+ *         if (!this.vizRenderer.has(renderMethod)) {
+ *           throw new Error(`Render method '${renderMethod}' has not been registered`);
+ *         }
+ *
+ *         return this.vizRenderers.get(renderMethod)(domElement);
```

Comment:
**nit** I had to read carefully to understand _where_ the handler is called. may I propose to be more explicit?
```diff
- return this.vizRenderers.get(renderMethod)(domElement);
+ const handler = this.vizRenderers.get(renderMethod);
+ return handler(domElement);
```